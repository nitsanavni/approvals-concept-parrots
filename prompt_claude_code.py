import subprocess
from filecache import filecache  # type: ignore[import-untyped]


@filecache
def prompt_claude_code_cached(prompt: str) -> str:
    """
    Call Claude Code via bunx with the given prompt (cached version).

    Args:
        prompt: The prompt to send to Claude Code

    Returns:
        The output from Claude Code as a string
    """
    cmd = ["bunx", "--bun", "claude", "--dangerously-skip-permissions", "-p", prompt]

    result = subprocess.run(cmd, capture_output=True, text=True, check=True)

    return result.stdout


def prompt_claude_code(prompt: str) -> str:
    """
    Wrapper function that calls the cached version.

    Args:
        prompt: The prompt to send to Claude Code

    Returns:
        The output from Claude Code as a string
    """
    return prompt_claude_code_cached(prompt)
