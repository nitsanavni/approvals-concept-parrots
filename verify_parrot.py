#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.12"
# dependencies = ["approvaltests", "pickle"]
# ///

import pickle
import inspect
import json
from typing import List, Callable, Any
from pathlib import Path
from approvaltests import verify, Namer


def verify_parrot(fn: Callable, args: List[List]) -> None:
    fn_module = inspect.getmodule(fn)
    fn_file = inspect.getfile(fn) if fn_module else "unknown"
    fn_name = fn.__name__
    
    fn_file_path = Path(fn_file)
    fn_file_name = fn_file_path.stem if fn_file_path.name != "unknown" else "unknown"
    
    cache_data = {
        "function": f"{fn_file_name}.{fn_name}",
        "file": str(fn_file),
        "results": []
    }
    
    for arg_list in args:
        try:
            result = fn(*arg_list)
            cache_data["results"].append({
                "args": arg_list,
                "result": result
            })
        except Exception as e:
            cache_data["results"].append({
                "args": arg_list,
                "error": str(e)
            })
    
    # Save pickle file
    pickle_file = Path(f"{fn_file_name}-{fn_name}.pickle")
    with open(pickle_file, 'wb') as f:
        pickle.dump(cache_data, f, protocol=pickle.HIGHEST_PROTOCOL)
    
    # Create human-readable format for approval testing
    human_readable = f"Function: {cache_data['function']}\n"
    human_readable += f"File: {cache_data['file']}\n"
    human_readable += "-" * 50 + "\n"
    
    for entry in cache_data['results']:
        if 'error' in entry:
            human_readable += f"Args: {entry['args']} -> ERROR: {entry['error']}\n"
        else:
            human_readable += f"Args: {entry['args']} -> Result: {entry['result']}\n"
    
    class CustomNamer(Namer):
        def get_approved_filename(self, base=None):
            return f"{fn_file_name}-{fn_name}.approved.txt"
        
        def get_received_filename(self, base=None):
            return f"{fn_file_name}-{fn_name}.received.txt"
    
    verify(human_readable, namer=CustomNamer())