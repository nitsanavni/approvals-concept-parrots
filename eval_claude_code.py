import subprocess

from prompt_claude_code import prompt_claude_code


def create_code_only_prompt(intention_prompt: str) -> str:
    """
    Create a prompt that instructs Claude to return only valid Python code.
    
    Args:
        intention_prompt: Description of what the Python code should do
        
    Returns:
        Formatted prompt string for Claude
    """
    return f"""Return ONLY valid Python code with no explanations, no markdown, no decorations.
Just the raw Python code that accomplishes this:
{intention_prompt}

Important: Return ONLY executable Python code, nothing else."""


def eval_claude_code(intention_prompt: str) -> str:
    """
    Prompt Claude to generate Python code for the given intention and evaluate it.
    
    Args:
        intention_prompt: Description of what the Python code should do
        
    Returns:
        String containing both the generated code and its execution result
    """
    # Get code from Claude
    code_prompt = create_code_only_prompt(intention_prompt)
    code = prompt_claude_code(code_prompt).strip()
    
    # Remove any markdown code blocks if they somehow appear
    if code.startswith("```"):
        lines = code.split("\n")
        # Find start and end of code block
        code_lines = []
        in_block = False
        for line in lines:
            if line.startswith("```") and not in_block:
                in_block = True
                continue
            elif line.startswith("```") and in_block:
                break
            elif in_block:
                code_lines.append(line)
        code = "\n".join(code_lines)
    
    # Execute the code using subprocess
    try:
        result = subprocess.run(
            ["uv", "run", "python", "-c", code],
            capture_output=True,
            text=True,
            check=True
        )
        result_str = f"Code:\n{code}\n\nResult: {result.stdout.strip() if result.stdout else 'None'}"
    except subprocess.CalledProcessError as e:
        result_str = f"Code:\n{code}\n\nError: {e.stderr if e.stderr else str(e)}"
    except Exception as e:
        result_str = f"Code:\n{code}\n\nError: {e}"
    
    return result_str