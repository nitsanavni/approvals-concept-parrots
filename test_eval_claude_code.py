from approvaltests import verify
from eval_claude_code import eval_claude_code
from verify_parrot import parrot
from prompt_claude_code import prompt_claude_code


def test_eval_claude_code_hello_world():
    with parrot(prompt_claude_code):
        verify(
            eval_claude_code(
                "Write a function that returns the string 'Hello, World!' and call it"
            )
        )


def test_eval_claude_code_calculator():
    with parrot(prompt_claude_code):
        verify(
            eval_claude_code("Create a simple calculator that adds two numbers: 5 + 3")
        )


def test_eval_claude_code_list_comprehension():
    with parrot(prompt_claude_code):
        verify(
            eval_claude_code(
                "Generate a list comprehension that squares numbers from 1 to 5"
            )
        )


def test_eval_claude_code_fibonacci():
    with parrot(prompt_claude_code):
        verify(eval_claude_code("Print the first 10 Fibonacci numbers"))


def test_eval_claude_code_dictionary():
    with parrot(prompt_claude_code):
        verify(
            eval_claude_code(
                "Create a dictionary with keys 'a', 'b', 'c' and values 1, 2, 3, then print it"
            )
        )


def test_eval_claude_code_factorial():
    with parrot(prompt_claude_code):
        verify(eval_claude_code("Calculate the factorial of 5"))
