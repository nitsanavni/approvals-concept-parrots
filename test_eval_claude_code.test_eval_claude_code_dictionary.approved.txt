Code:
dictionary = {'a': 1, 'b': 2, 'c': 3}
print(dictionary)

Result: {'a': 1, 'b': 2, 'c': 3}
